cmd/kubeadm/app/util/endpoint.go

```diff
@@ -37,15 +38,21 @@ func GetMasterEndpoint(cfg *kubeadmapi.MasterConfiguration) (string, error) {
 // GetMasterHostPort returns a properly formatted Master IP/port pair or error
 // if the IP address can not be parsed or port is outside the valid TCP range.
 func GetMasterHostPort(cfg *kubeadmapi.MasterConfiguration) (string, error) {
-	masterIP := net.ParseIP(cfg.API.AdvertiseAddress)
-	if masterIP == nil {
-		return "", fmt.Errorf("error parsing address %s", cfg.API.AdvertiseAddress)
+	var masterIP string
+	if len(cfg.API.ControlPlaneEndpoint) > 0 {
+		masterIP = cfg.API.ControlPlaneEndpoint
```

Comment:
`validation.IsDNS1123Subdomain(ControlPlaneEndpoint)` here as well, and if the validation fails, error out.